### ISSUE TITLE
false-positive: unreachable pattern

### ISSUE THREAD

**Issue by MrGibus** [intellij-rust/intellij-rust#8634]
## Environment

* **IntelliJ Rust plugin version: 0.4.165.4438-213**
* **Rust toolchain version: 1.52.1 stable-x86_64-pc-windows-msvc**
* **IDE name and version: PyCharm 2021.3.2 (Community Edition) Build #PC-213.6777.50, built on January 28, 2022**
* **Operating system: Windows 10**

## Problem description
Updated to 2021.3.2 and had many false positives where there were none before. Managed to remove most of them by invalidating caches but this one persists.
I've searched "is:issue false positive unreachable" but nothing turned up.

## Steps to reproduce

Make a new project and paste this:
```rust
// There's bugs in this it's because I crudely pasted it into a new project and stripped out all the functionality
// Conventions follow a mix of WinAPI conventions and rust conventions. please ignore.
// The real implementation works as intended.
// IDE still reports that all but the first pattern is unreachable even before adding dependencies and imports
// Weirdly IDE does not report that it doesn't know what these values are, even though I haven't imported anything.
unsafe fn event_handler(
    hwnd: HWND,
    message: u32,
    wparam: WPARAM,
    lparam: LPARAM,
) -> LRESULT {
    match message{
            WM_NCCALCSIZE => {
                todo!()
            }
            WM_PAINT => {
                todo!()
            }
            WM_NCHITTEST => {
                todo!()
            },
            WM_DESTROY => {
                todo!()
            }
            _ => DefWindowProcW(hwnd, message, wparam, lparam)
    }
}
```

**Comment by denkuy** [comment]
Because it doesn't know about `WM_NCCALCSIZE` it just assumes that you want a variable called like this.

Running this will always assign `message`'s value to a variable called `WM_NCCALCSIZE` and run the first arm, hence the warning.

**Comment by MrGibus** [comment]
@denkuy That's pretty worrying behaviour, I thought variables would only assign to an enum. This does however, give a little insight to the issue-- sounds less to do with unreachable patterns and more to do with the constants in the crate not being understood in the plugin. These constants originate in the windows-rs crate.

**Comment by afetisov** [comment]
> @MrGibus That's pretty worrying behaviour, I thought variables would only assign to an enum.

Note that this behaviour is mandated by rustc rather than the plugin. If Rust sees an identifier in a pattern and it's not a known constant, then it is always a binding of a new variable. This is the main reason why constants are written in ALL_CAPS --- to make it clear whether the value is intended to be bound, and to avoid accidental name collisions between bindings and constants.
